src/dynamic_programming/maximum_subarray.rs

```diff
@@ -1,62 +1,73 @@
 
-/// maximum_subarray(array) find the subarray (containing at least one number) which has the largest sum
-/// and return its sum.
+/// Finds the subarray (containing at least one number) which has the largest sum
+/// and returns its sum.
 ///
 /// A subarray is a contiguous part of an array.
 ///
-/// Arguments:
-///     * `array` - an integer array
-/// Complexity
-///     - time complexity: O(array.length),
-///     - space complexity: O(array.length),
-pub fn maximum_subarray(array: &[i32]) -> i32 {
-    let mut dp = vec![0; array.len()];
-    dp[0] = array[0];
-    let mut result = dp[0];
-
-    for i in 1..array.len() {
-        if dp[i - 1] > 0 {
-            dp[i] = dp[i - 1] + array[i];
-        } else {
-            dp[i] = array[i];
-        }
-        result = result.max(dp[i]);
+/// # Arguments
+///
+/// * `array` - A slice of integers.
+///
+/// # Returns
+///
+/// An integer representing the largest sum of a contiguous subarray.
+///
+/// # Complexity
+///
+/// * Time complexity: O(array.len())
+/// * Space complexity: O(array.len())
+pub fn maximum_subarray(array: &[isize]) -> isize {
+    if array.is_empty() {
+        return 0;
     }
 
-    result
+    let mut max_sum = vec![0; array.len()];
+    max_sum[0] = array[0];
+
+    array.iter().enumerate().skip(1).for_each(|(i, &x)| {
+        max_sum[i] = (max_sum[i - 1] + x).max(x);
+    });
+
+    *max_sum.iter().max().unwrap()
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
 
-    #[test]
-    fn non_negative() {
-        //the maximum value: 1 + 0 + 5 + 8 = 14
-        let array = vec![1, 0, 5, 8];
-        assert_eq!(maximum_subarray(&array), 14);
-    }
-
-    #[test]
-    fn negative() {
-        //the maximum value: -1
-        let array = vec![-3, -1, -8, -2];
-        assert_eq!(maximum_subarray(&array), -1);
-    }
-
-    #[test]
-    fn normal() {
-        //the maximum value: 3 + (-2) + 5 = 6
-        let array = vec![-4, 3, -2, 5, -8];
-        assert_eq!(maximum_subarray(&array), 6);
+    macro_rules! maximum_subarray_tests {
+        ($($name:ident: $tc:expr,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (array, expected) = $tc;
+                    assert_eq!(maximum_subarray(&array), expected);
+                }
+            )*
+        }
     }
 
-    #[test]
-    fn single_element() {
-        let array = vec![6];
-        assert_eq!(maximum_subarray(&array), 6);
-        let array = vec![-6];
-        assert_eq!(maximum_subarray(&array), -6);
+    maximum_subarray_tests! {
+        test_all_non_negative: (vec![1, 0, 5, 8], 14),
+        test_all_negative: (vec![-3, -1, -8, -2], -1),
+        test_mixed_negative_and_positive: (vec![-4, 3, -2, 5, -8], 6),
+        test_single_element_positive: (vec![6], 6),
+        test_single_element_negative: (vec![-6], -6),
+        test_mixed_elements: (vec![-2, 1, -3, 4, -1, 2, 1, -5, 4], 6),
+        test_empty_array: (vec![], 0),
```

Comment:
According to [line 5](https://github.com/TheAlgorithms/Rust/blob/18fe64aeb95674f96aef5de30f46771d4a5188ee/src/dynamic_programming/maximum_subarray.rs#L5), this should return some error.

Reply:
Added custom error type